From b41173bd5f11e86577b36b138ded6f815977fe9e Mon Sep 17 00:00:00 2001
From: Victor Kamensky <kamensky@cisco.com>
Date: Sat, 23 Dec 2017 05:14:06 -0600
Subject: [PATCH 14/15] selinux: allow setxattr on rootfs so initramfs code can
 set them

initramfs code supporting extended cpio format have ability to
fill extended attributes from cpio archive, but if SELinux enabled
and security server is not initialized yet, selinux callback would
refuse setxattr made by initramfs code.

Solution enable SBLABEL_MNT on rootfs even if secrurity server is
not initialized yet.

Signed-off-by: Victor Kamensky <kamensky@cisco.com>
---
 security/selinux/hooks.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/security/selinux/hooks.c b/security/selinux/hooks.c
index 819fd6858b49..5797bc3e4b37 100644
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c
@@ -701,6 +701,17 @@ static int selinux_set_mnt_opts(struct super_block *sb,
 
 	if (!ss_initialized) {
 		if (!num_opts) {
+			/*
+			 * Special handling for rootfs. Is genfs but supports
+			 * setting SELinux context on in-core inodes.
+                         *
+                         * Chicken and egg problem: policy may reside in rootfs
+                         * but for initramfs code to fill in attributes, it needs
+                         * selinux to allow that.
+			 */
+			if (strncmp(sb->s_type->name, "rootfs", sizeof("rootfs")) == 0)
+			    sbsec->flags |= SBLABEL_MNT;
+
 			/* Defer initialization until selinux_complete_init,
 			   after the initial policy is loaded and the security
 			   server is ready to handle calls. */
-- 
2.14.3

